diff --git a/meson.build b/meson.build
index e327498f..6ceab4c0 100644
--- a/meson.build
+++ b/meson.build
@@ -4,7 +4,7 @@ project(
     license: 'GPL 2+',
     # We'd like to use version: files('RECOLL-VERSION.txt') but this is a 0.56 feature. Buster has
     # 0.49 and focal 0.53. let's support at least focal backports for now.
-    version: '1.38.0pre1',
+    version: '1.38.0',
     default_options: ['cpp_std=c++17', 'buildtype=debugoptimized'],
     meson_version: '>=0.49',
 )
@@ -17,13 +17,16 @@ libxml = dependency('libxml-2.0')
 libxslt =  dependency('libxslt')
 libz = dependency('zlib')
 if get_option('libmagic')
-    libmagic = dependency('libmagic')
+    libmagic = dependency('libmagic', required:false)
+    # Libmagic has no pkgconfig data on older debian systems
+    if not libmagic.found()
+        libmagic = cc.find_library('libmagic', required:true)
+    endif
 endif
 
 # Check if anything needs Xapian. We also need to build the shared lib if this is the case.
 # The only case where it is not needed is for building rclgrep alone.
-if get_option('testmains') or get_option('python-module') or get_option('indexer') \
-   or get_option('xadump') or get_option('qtgui') or get_option('recollq')
+if get_option('testmains') or get_option('python-module') or get_option('indexer') or get_option('xadump') or get_option('qtgui') or get_option('recollq')
     librecoll_needed = true
     librecolldeps = []
     xapian = dependency('xapian-core')
