diff --git a/qtgui/rclmain_w.cpp b/qtgui/rclmain_w.cpp
index 0b93dcfa..549f0d5f 100644
--- a/qtgui/rclmain_w.cpp
+++ b/qtgui/rclmain_w.cpp
@@ -737,7 +737,14 @@ void RclMain::fileExit()
     rwSettings(true);
 
     deleteAllTempFiles();
+
+#if (QT_VERSION < QT_VERSION_CHECK(6, 0, 0)) && !defined(PREVIEW_FORCETEXTBROWSER)
+    // For some reason, qt5 qApp->exit(0) here stops desktop shutdown (at least on KDE and
+    // Cinnamon). Recoll exits but the shutdown is aborted.
+    exit(0);
+#else
     qApp->exit(0);
+#endif
 }
 
 // Start a db query and set the reslist docsource
