diff --git a/packaging/rpm/recoll.spec b/packaging/rpm/recoll.spec
index 62cac294..1a1ea621 100644
--- a/packaging/rpm/recoll.spec
+++ b/packaging/rpm/recoll.spec
@@ -10,6 +10,7 @@ License:        GPLv2+
 URL:            http://www.lesbonscomptes.com/recoll/
 Source0:        http://www.lesbonscomptes.com/recoll/recoll-%{version}.tar.gz
 Source10:       qmake-qt5.sh
+Patch0:         recoll-searchgui-desktop-encoding.patch
 BuildRequires:  aspell-devel
 BuildRequires:  bison
 BuildRequires:  desktop-file-utils
@@ -23,6 +24,8 @@ BuildRequires:  python2-devel
 BuildRequires:  python3-devel
 BuildRequires:  xapian-core-devel
 BuildRequires:  zlib-devel
+BuildRequires:  chmlib-devel
+BuildRequires:  libxslt-devel
 Requires:       xdg-utils
 
 %description
@@ -43,6 +46,7 @@ displayed in Konqueror.
 
 %prep
 %setup -q -n %{name}-%{version}
+%patch0 -p2
 
 %build
 CFLAGS="%{optflags}"; export CFLAGS
diff --git a/src/desktop/recoll-searchgui.desktop b/src/desktop/recoll-searchgui.desktop
index 3cceace7..bcd3ccf3 100644
--- a/src/desktop/recoll-searchgui.desktop
+++ b/src/desktop/recoll-searchgui.desktop
@@ -1,10 +1,10 @@
 [Desktop Entry]
 Categories=Qt;Utility;Filesystem;Database;
 Comment=Find documents by specifying search terms
-Comment[ru]=Поиск документов по заданным условиям
+Comment[ru]=п÷п╬п╦я│п╨ п╢п╬п╨я┐п╪п╣п╫я┌п╬п╡ п©п╬ п╥п╟п╢п╟п╫п╫я▀п╪ я┐я│п╩п╬п╡п╦я▐п╪
 Exec=recoll
 GenericName=Local Text Search
-GenericName[ru]=Локальный текстовый поиск
+GenericName[ru]=п⌡п╬п╨п╟п╩я▄п╫я▀п╧ я┌п╣п╨я│я┌п╬п╡я▀п╧ п©п╬п╦я│п╨
 Icon=recoll
 Name=Recoll
 Terminal=false
